Migrate react-scripts to vite

mfechner opened the pull request:
Fixes https://github.com/gotify/server/issues/695

To test this modification:
```
(cd ui && yarn)
(cd ui && yarn build)
go run .
(cd ui && yarn start)

```
Now access the address `yarn start `shows you.

mfechner commented:
Ok, to sum it up.
The migration to vite is more or less done.
But it seems that state management is broken, after a login the message component is not loaded.

I tried to update the router, but that is even worse. They broke the complete API and do not even provide a migration script for it.

As I mainly develop in Angular I need to read about react before I continue, but it seems that this upgrade to vite is depending on updating all libs to a more current version.

If someone else here have a strategy how to update the ui, any feedback is welcome.
I understand now the comment from @jmattheis  that this will not be an easy task ;)

jmattheis commented:
@mfechner I feel you (:, I think the only good way it to start from scratch and copy component by component into a new project, so that you don't have to fix 100 errors to get something working. It's on my todo list, sadly it's multi hour effort.

mfechner commented:
@jmattheis thanks a lot for your feedback. I currently study react to better understand react itself.
It's to early to really make a statement here, but why you do not use the standard state management of react. That would help to get rid of a third party lib and would also simplify upgrade in the future.

I'm on this task but it will take some time ;)

jmattheis commented:
The codebase is quite old, pre react hooks. In my experience it wasn't that ergonomic to use contexts for global state, so I've used a lib. Currently I like https://github.com/pmndrs/zustand for it's simplicity and small footprint.

mfechner commented:
I started now with a rewrite, but this will be a massiv change.
Steps I plan to do:

- rewrite all class based components as functional components to be able to use all hook
- upgrade react in first step to 18.3.1 and later to 19
- upgrade material to v6
- replace state management with `@reduxjs/toolkit` and `react-redux`
- upgrade react-router to 7
- use vite 6

I think it will be nearly impossible to get this changes sliced into small commit, as all steps link somehow together.
Maybe later try to get also rid of not needed third party libs like axios. The fetch API is normally more than enough nowadays.

I update you, if I have something to show, but it will for sure take some days maybe some weeks.

mfechner commented:
This is now a first version we can start.
The newly rewritten react app looks ok and tests I did are working fine.
There can be some finetuning, like loading indicator or use redux RTK query.

There are some todos left (infinite loading component for messages), but it is now a good starting point.

The only problem I have is that puppeteer sometimes is just not loading the page and therefore tests are not running.
Maybe someone has an idea why?

@jmattheis would be nice if you can have a look, it was really a lot of work

mfechner commented:
All tests are now working fine for me.

jmattheis commented:
I'll have a look at this in the coming days, but I'll take some time.

mfechner commented:
> I'll have a look at this in the coming days, but I'll take some time.

thank you very much. I found other small issues I corrected in the meantime.
The webinterface itself seems to work, but I have problem with the tests.
Sometimes they are failing (about 2 failures if doing 10 test runs).

The problems I found is that for most actions a submit or a click on a button was not blocking (running in background), which I reverted to the behaviour the released gotify version has.
But something with the test is maybe broken and as I never used pupeteer before, that is hard for me to analyse, so I need some help with the tests.

Later we can addnice indicators, if you e.g. click on Submit the button is replaced by an loading indicator.

I have here some very nice ideas, but I will provide this later in another PR, as that is not related to this topic.

Let us first get the update to react 19 merged, then we can continue further in smaller steps.

Maybe we can release these changes as a beta version to enable users testing it.

jmattheis commented on ui/src/message/Messages.tsx:
```diff
@@ -1,26 +1,118 @@
+                <LoadingSpinner />
+            ) : hasMessages ? (
+                <div style={{width: '100%'}} id="messages">
+                    {/* TODO: maybe replace ReactInfitite with react-window, which is also documented here: https://mui.com/material-ui/react-list/#virtualized-list */}
```
I've recently tried a lot of window libraries for another project and landed on https://tanstack.com/virtual/latest/docs/introduction react-window doesn't play to well with having a window over the whole screen.

jmattheis commented on ui/src/common/SettingsDialog.tsx:
```diff
@@ -1,68 +1,64 @@
-class SettingsDialog extends Component<IProps & Stores<'currentUser'>> {
-    @observable
-    private pass = '';
+// TODO: rename to ChangePasswordDialog
```
Please make as few refactoring as possible in this PR, only what's necessary so it says small.

mfechner replied:
ok, I removed that todo again and we keep the component as a function component

jmattheis commented on ui/src/store/ui-slice.ts:
```diff
@@ -0,0 +1,63 @@
+}
+
+const initialUiState: UiState = {
+    themeKey: 'dark',
```
I'm not sure I like this too, previously this was local component state, because it was only used in two places, and passed downwards to one component. Sure adding this to the global state can be done, but local component state is better as it's simpler and easier to navigate.

These refactorings making this review much more effort as it's not just library change / upgrades, but also logic changes which aren't 100% necessary.

mfechner replied:
I had a lot of problems in the beginning to understand which prop is passed to which component and why
Using a global state reduces the need of props drilling which is making the componets match leaner and easier to read.

But if you prefer a global state in e.g. app.tsx, I can remove these UI related states from the global state and move it to app.tsx and pass them via props drilling to the components.
But to be honest it will be technically equal and will just make the components harder to read.

But as you prefer.

jmattheis replied:
I prefer it to be as close as possible to the original implementation for now, as it makes less changes to review and less possibility of bugs, because the code already runs for some time. This includes the moving of the selected plugin state into redux.

mfechner replied:
ok, I will move the state/slice/action definition to the component.
But I will not remove the ui state management and will change everything back to prop drilling.
This is error prone and we will for sure just add new bugs by this. The UI state is working perfectly fine.

jmattheis replied:
> This is error prone and we will for sure just add new bugs by this. 

How is it error prone to pass properties to sub components. TypeScript is used, so it's not like you can miss passing the props?

jmattheis reviewed (commented):
Not finished with reviewing, this was only glancing through some files.

mfechner commented on ui/src/App.tsx:
```diff
@@ -0,0 +1,114 @@
+import React, {useEffect} from 'react';
+import {createBrowserRouter, RouterProvider} from 'react-router-dom';
+import Applications from './application/Applications.tsx';
```
Yes, Material changed the color schema with the upgrade, I remember a note in the migration guide. I will see how to change this back to previous color for the dark theme.

mfechner replied:
Is fixed, will commit it soon.

mfechner commented on ui/src/App.tsx:
```diff
@@ -0,0 +1,114 @@
+import React, {useEffect} from 'react';
```
Yes, was caused by changes in Material upgrade, will commit the fix soon.

mfechner commented on ui/src/App.tsx:
```diff
@@ -0,0 +1,114 @@
+import React, {useEffect} from 'react';
+import {createBrowserRouter, RouterProvider} from 'react-router-dom';
```
I think this is also fixed now, will commit it soon.

mfechner commented on ui/index.html:
```diff
@@ -35,8 +35,6 @@
     Gotify requires JavaScript.
 </noscript>
 <div id="root"></div>
-<% if (process.env.NODE_ENV === 'production') { %>
-    <script>window.config = %CONFIG%;</script>
-<% } %>
```
Thanks a lot, is now also fixed

mfechner commented on ui/package.json:
```diff
@@ -5,69 +5,70 @@
   "homepage": ".",
   "proxy": "http://localhost:80",
```
Is fixed, will commit it soon.

mfechner commented on ui/src/store/app-actions.ts:
```diff
@@ -0,0 +1,104 @@
+        if (!getAuthToken()) {
+            return;
+        }
+        const sendRequest = async () => {
+            const response = await axios.get<IApplication[]>(`${config.get('url')}application`);
+
+            return response.data;
+        };
+
+        // TODO: handle error case
+        const appData = await sendRequest();
```
I have written it like this with the idea to remove axios from the bundle and switch to standard fetch API.
Writting it like this is easier to catch errors and rethrow custom errors if necessary.

But if you to stick to axios, I can rewrite all queries.
@jmattheis please let me know if you want to stick to axios or if you want to get rid of third party axios module

jmattheis replied:
Yeah, fetch should be used instead, but migration/removal should be done in a separate PR.

jmattheis replied:
I'm not sure I understand the reasoning, currently there is no error handling but your expecting there to be some? There is probably some utility function that calls fetch with the error handling, so I don't think there is some special error handling for specific calls (besides logging in).

mfechner replied:
ok, I removed that pattern. Please do not squash the commit while merging this PR, so I can undo that commit later if we switch to fetch API.

mfechner commented on ui/src/store/app-actions.ts:
```diff
@@ -0,0 +1,104 @@
+import axios from 'axios';
```
I have written it like this, because now it is generic and simple.
The idea to switch in a second step to RTK-Query:
https://redux-toolkit.js.org/rtk-query/overview

which would also provide a loading state and more and can be rewritten with less code, but is then bound to a specific state library.
Like it is written now, it should be very simple to switch to any state library without any rewrite required in code.

And if you prefer to have the state not in a common state directory, I can move them back into the components.
I prefer to have the state in one common directory, but having then on the component is also fully valid and a matter of preference.

@jmattheis please let me know, what you prefer, I personally do not care ;)

jmattheis replied:
Have you tried moving to a newer mobx version before migrating to redux? I'd prefer this for this PR, as it would reduce logic changes in this migration PR.

mfechner replied:
I do not know mobx, but what I understand is that it is used in combination with class based components.
But as we now moved away from class based components, I'm not sure if using it is a good idea.

My problem is, I will not have the time to read through a new framework.

Do you not want to give redux-toolkit a try?
I think in combination with queryRTK it is a good decision and we can reduce the boilerplate with it significantly.

jmattheis replied:
The newer version of mobx supports/encourages functional components.

> My problem is, I will not have the time to read through a new framework.

same for me with redux/redux-toolkit

> Do you not want to give redux-toolkit a try?

Maybe, but I dislike this format, this is a migrate build system PR, not rewrite whole statemanagement and also do updates alongside it. I feel like it should be possible to do the updates with minimal changes to the actual code, and then do rewrites like this separately in smaller batches so they can be discussed accordingly.
 
Each change, be it the added loading spinners, add additional complexity to this PR, that make bugs more likely to through, as it makes this PR even bigger. It may look like an changes that is positive in all ways, but as a reviewer I'll still question every change like this if it's reasonable.

I do _dislike_ mobx quite a bit, but I know, that the code is mostly bug free because it already ran for multiple years. So this basically trades mature code with a re implementation without big benefits.

It's also not clear to me that redux should be used instead, I'm not 100% up-to-date on the react state management libs, and given I'll have to maintain this afterwards I don't want to just accept a new state management lib from a contribution without any own research if it fits.

But yeah, this is again for effort for a change that is already quite big.

mfechner commented on ui/src/store/app-actions.ts:
```diff
@@ -0,0 +1,104 @@
+
+export const uploadImage = (id: number, file: Blob) => {
+    return async (dispatch: AppDispatch) => {
+        const sendRequest = async () => {
+            const response = await axios.post(
+                `${config.get('url')}application/${id}/image`,
+                formData,
+                {
+                    headers: {'content-type': 'multipart/form-data'},
+                }
+            );
+            return response.data;
+        };
+
+        const formData = new FormData();
+        formData.append('file', file);
```
as explained before, I use that pattern as I try to use as less third party modules and stick the core modules.
But if you want to stick to axios, I will rewrite all actions to be as simple as possible.

mfechner commented on ui/src/store/app-actions.ts:
```diff
@@ -0,0 +1,104 @@
+
+        // TODO: handle error case
+        await sendRequest();
+        dispatch(appActions.remove(id));
```
But this will cause a slower user interface for the user as the browser is sending a DELETE request, get a confirmation and is refetching the data again, which is not necessary.

I tried to get all actions to behave equally (messages, app, clients, ...), so I can switch it later to a more high level implementation which uses less boilerplate code like RTKQuery of tanstack.

mfechner commented on ui/vite.config.ts:
```diff
@@ -0,0 +1,102 @@
+        outDir: 'build',
+        assetsDir: 'static',
+        emptyOutDir: true,
+        sourcemap: 'inline',
```
Good catch, I had that in while debugging, as the app now is mostly running, I will remove it again.

mfechner commented on ui/vite.config.ts:
```diff
@@ -0,0 +1,102 @@
+                ws: true,
+                rewrite: (p) => p.replace(/^\/api/, ''),
+                rewriteWsOrigin: true,
+            }
```
I think you tested an older version, that should be fixed already by:
https://github.com/gotify/server/pull/696/commits/0e3529f89d1e38c60c2192bdc104b5d440bbcb14

jmattheis replied:
Yeah, it works now :+1:

mfechner commented on ui/src/store/message-slice.ts:
```diff
@@ -0,0 +1,47 @@
+import {IMessage, IPagedMessages} from '../types.ts';
+
+interface MessagesState {
+    items: IMessage[];
```
ok, I will add this as a TODO to me list

mfechner replied:
Already fetch messages (by appId) are now cached in state and a refresh of a single app is now not removing other messages anymore.

mfechner commented on ui/vite.config.ts:
```diff
@@ -0,0 +1,102 @@
+    server: {
+        host: '0.0.0.0',
+        proxy: {
+            '/api': {
```
I simplified the approach. We have now only one url that is used and vite is taking care to forward API requests now not by `/api` prefix but by the path selector you suggested. I added also the path `image` to be able to load app images successfully.

mfechner commented on ui/src/message/Messages.tsx:
```diff
@@ -1,26 +1,118 @@
+    const hasMessages = messages.length !== 0;
 
+    useEffect(() => {
+        dispatch(fetchMessages(appId));
```
I think that should be fixed now.

mfechner replied:
I think there is a race condition.
I added now loading indicators.

mfechner commented on ui/serve.go:
```diff
@@ -32,7 +31,7 @@ func Register(r *gin.Engine, version model.VersionInfo, register bool) {
 		return strings.Replace(content, "%CONFIG%", string(uiConfigBytes), 1)
 	}
 
-	ui := r.Group("/", gzip.Gzip(gzip.DefaultCompression))
```
If a test is failing, I always see the following error message in the javascript of the puppeteer browser:

![grafik](https://github.com/user-attachments/assets/6842423f-f9ba-4f8a-aafb-459bd8a3e5dc)

or

![grafik](https://github.com/user-attachments/assets/00c7079d-c4d0-445a-981b-e70b03c23dcd)

Not sure if that is a problem related to my machine, some virus scanner or anything else.

But if I remove this commit again, all tests are failing 100% with the error message:
![grafik](https://github.com/user-attachments/assets/225cfefc-b586-4f41-bfde-62675324b0c6)

mfechner replied:
I removed it from this PR:
[Fixed error net:ERR_INCOMPLETE_CHUNKED_ENCODING](https://github.com/mfechner/gotify-server/commit/4b088d091dc2cfe74bb47e62d776349bd85ac52f) in branch https://github.com/mfechner/gotify-server/commits/fix_err_incomplete_chunked_encoding/

mfechner commented:
Failing tests seem to be related to puppeteer upgrade.
I downgraded it and I see the tests are working better.

I will work through the commits again and will maybe rebase some of them, so they match better.

@jmattheis the current implementation does not have the virtual scrolling for messages, should I include it here, ot as another PR?
Also paging is currently not implemented (which I can currently not test, so I have to create a datasource to work with paging).

jmattheis commented:
Depending on my decision in https://github.com/gotify/server/pull/696#discussion_r1906019369 I'd want the same functionality as on current master in this PR.

jmattheis commented on ui/src/reactions.ts:
```diff
@@ -1,46 +0,0 @@
-        }
-    );
-
-    reaction(
-        () => stores.currentUser.connectionErrorMessage,
-        (connectionErrorMessage) => {
-            if (!connectionErrorMessage) {
-                clearAll();
-                loadAll();
-            }
-        }
-    );
```
This seems to be not migrated into redux.

mfechner replied:
Could you please explain what this code is doing or better what is the propose for it?

I do not see a reason, why you want to destroy your complete local store if the server is not available.
The store is destroyed if the user got invalid.

jmattheis replied:
If the client loses connection, then the client doesn't receive new messages via the websocket connection. When the connection is reestablished, we have to refetch everything to get an up-to-date state.

jmattheis commented on ui/src/App.tsx:
```diff
@@ -0,0 +1,115 @@
+    const loggedIn = useAppSelector((state) => state.auth.loggedIn);
+
+    useEffect(() => {
+        if (loggedIn) {
```
Is this really the way to react to changes in the state, it there a native way in redux to subscribe to changes in the state and do stuff? Doing this in react seems weird, and having the global state besides redux store (the websocket store varibale in this case, or reconnectTimeoutId in the auth-actions.ts) seems weird to. The redux actions shouldn't have side effects with other global variables.

mfechner replied:
Is is native react, see also here:
https://react.dev/reference/react/useEffect

useEffect is called if the component is rendered and is called again if one of the dependencies are changing.
This hook is ment to be used to react to any changes that are relevant.

And as App.tsx is a very high level root component I thought it is a good place to initiate the websocket connection or close it depending on login status of the user.

I do not know of a mechanism in the state itself, as that would be redunant to the hook.

jmattheis replied:
Yeah, it does work. I see useEffect more as a side effect for this specific component, so synchronizing the component state with a side effect. 

This useEffect usage does only manage a global side effect for the global store, so it basically doesn't need to have a react dependency. Additionally this will unnecessarily rerender the component when the loggedIn state changes, only to have the useEffect executed.

Other state management libs have this e.g.
- reaction from mobx
- zustand subscribe https://github.com/pmndrs/zustand/tree/main?tab=readme-ov-file#readingwriting-state-and-reacting-to-changes-outside-of-components
- jotai subscribe https://jotai.org/docs/core/store

The rerender is not a problem in this case, as this isn't a complex application but I had expected that redux has something similar.

mfechner replied:
Is committed.

jmattheis commented on ui/src/application/Applications.tsx:
```diff
@@ -1,64 +1,94 @@
+    const [createDialog, setCreateDialog] = useState<boolean>(false);
+
+    const fileInputRef = useRef<HTMLInputElement>(null);
+    let uploadId = -1;
```
Not sure why this works, but this should probably be a React.useRef(), otherwise it will be overriden by a rerender.

mfechner replied:
The uploadId is used to store the ID, the document you upload, belongs to.
In this case the appId, which is stored if you click on the upload button in the applications table:
![grafik](https://github.com/user-attachments/assets/68e14dc0-ae05-45a3-ba59-57e6ff1db388)

That would mean, that the component must redender while you have the browser native upload file selector window open. And that is not possible, as the file selector window is a modal against the browser window.
(we do not have any side effect in place that is reloading apps in background)

As that is a simple variable, we cannot use a ref for it.
But I can move that variable outside the component, then it would survive a rerender, but which is in that case not necessary, as we do not have any side effect in place that would trigger a rerender of the component while the file selector window is opened.

jmattheis replied:
> That would mean, that the component must redender while you have the browser native upload file selector window open. And that is not possible, as the file selector window is a modal against the browser window.
(we do not have any side effect in place that is reloading apps in background)

That it currently does not occur, doesn't mean it's impossible that it'll occur. Components should be designed, so that they don't break with future changes. Currently,  we already have global events like the useEffect in the App.tsx that opens the websocket connection. Stuff like this can cause a rerender of this component and prevent the successful uploading of the image. You can check this by adding this to the component:

```typescript
const [ignored, setUpdateNow] = useState(1);
useEffect(() => {
    const id = setInterval(() => setUpdateNow(i => i+1), 100);
    return () => clearInterval(id);
}, [])
```

> As that is a simple variable, we cannot use a ref for it.

I don't understand, it's a simple variables, therefore we cannot use a ref for it? The render problem can be fixed with the following change:

```diff
diff --git a/ui/src/application/Applications.tsx b/ui/src/application/Applications.tsx
index 4a138e2..c4d1b4b 100644
--- a/ui/src/application/Applications.tsx
+++ b/ui/src/application/Applications.tsx
@@ -33,7 +33,7 @@ const Applications = () => {
     const [createDialog, setCreateDialog] = useState<boolean>(false);

     const fileInputRef = useRef<HTMLInputElement>(null);
-    let uploadId = -1;
+    const uploadId = useRef(-1);
     // select app from global state

     // load applications from server
@@ -42,7 +42,7 @@ const Applications = () => {
     }, [dispatch]);

     const handleImageUploadClick = (id: number) => {
-        uploadId = id;
+        uploadId.current = id;
         if (fileInputRef.current) {
             fileInputRef.current.click();
         }
@@ -54,7 +54,7 @@ const Applications = () => {
             return;
         }
         if (['image/png', 'image/jpeg', 'image/gif'].indexOf(file.type) !== -1) {
-            dispatch(uploadImage(uploadId, file));
+            dispatch(uploadImage(uploadId.current, file));
         } else {
             alert('Uploaded file must be of type png, jpeg or gif.');
         }
```

> But I can move that variable outside the component,

No that's bad practice, and useRef is exactly made for this usecase.

mfechner replied:
Thanks @jmattheis, was not aware of that use-case for useRef, will commit a fix soon.

jmattheis commented on ui/src/user/auth-actions.ts:
```diff
@@ -0,0 +1,166 @@
+            .catch(() => Promise.resolve());
+
+        localStorage.removeItem(tokenKey);
+        dispatch(authActions.logout());
```
This naming is weird. In a file called auth-actions, we have a function called logout which uses authActions.logout from auth-slice.ts. Is this really the redux way? Having a filed called as a variable from another file which use each other seems complicated.

mfechner replied:
In the reducer you must not do any side effect, that is the reason, why the reducer and the action (sideeffect) is seperated, see also here:
https://redux.js.org/usage/migrating-to-modern-redux#reducers-and-actions-with-createslice

But yes, we can change the names.
Maybe use in the action doLogin and in the reducer loginDone or what naming you suggest?

mfechner commented on ui/package.json:
```diff
@@ -2,72 +2,71 @@
+    "@uiw/react-codemirror": "^4.23.7",
+    "axios": "1.7.9",
+    "detect-browser": "5.3.0",
+    "install": "^0.13.0",
```
I removed it

mfechner commented on ui/package.json:
```diff
@@ -2,72 +2,71 @@
+    "axios": "1.7.9",
+    "detect-browser": "5.3.0",
+    "install": "^0.13.0",
+    "js-base64": "3.7.7",
```
You are using it here:
https://github.com/gotify/server/blob/master/ui/src/CurrentUser.ts#L78

Should I rewrite this?

jmattheis replied:
Ahh your right, nvm then.

mfechner commented on ui/package.json:
```diff
@@ -2,72 +2,71 @@
+    "react-router": "7.1.1",
+    "react-router-dom": "7.1.1",
+    "react-timeago": "7.2.0",
+    "react-window": "1.8.11",
```
I removed it

mfechner commented:
> Depending on my decision in [#696 (comment)](https://github.com/gotify/server/pull/696#discussion_r1906019369) I'd want the same functionality as on current master in this PR. There is another open discussion which is unanswered.

ok, I will also commit other commits into this branch to restore functionality to match the currently release gotify version as good as possible.

jmattheis commented:
Hey, sorry for the long delay. I've created #818 that used parts of this PR but kept using mobx. This will hopefully be merged soon. Thanks for your work.
